etc/mpolicy/mpolicy.man
.\"
.\" Mach Operating System
.\" Copyright (c) 1991,1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\" 
.\" CARNEGIE MELLON ALLOWS FREE USE OF THIS SOFTWARE IN ITS 
.\" CONDITION.  CARNEGIE MELLON DISCLAIMS ANY LIABILITY OF ANY KIND FOR
.\" ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
.\" 
.\" Carnegie Mellon requests users of this software to return to
.\" 
.\"  Software Distribution Coordinator  or  Software.Distribution@CS.CMU.EDU
.\"  School of Computer Science
.\"  Carnegie Mellon University
.\"  Pittsburgh PA 15213-3890
.\" 
.\" any improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" HISTORY
.\" $Log:	mpolicy.man,v $
.\" Revision 2.3  91/03/19  12:31:48  mrt
.\" 	Changed to new copyright
.\" 
.\" Revision 2.2  90/10/08  13:15:54  rpd
.\" 	Created.
.\" 	[90/10/08  13:12:53  rpd]
.\" 
.TH MPOLICY 1 6/18/90
.CM 4
.SH NAME
mpolicy \- change scheduling policies
.SH SYNOPSIS
\fBmpolicy\fP [\fB-host\fP machine] [\fB-policy\fP policy data|\fB-enable\fP policy|\fB-disable\fP policy] [\fB-threads\fP] ids...
.SH DESCRIPTION
\fImpolicy(1)\fR may be used to change the scheduling policy of threads
and enable or disable scheduling policies for processor sets.
set.  The following scheduling policies are understood:
.TP 8
.B tshare
The time-sharing scheduling policy.  Recent CPU usage lowers the
priority of a thread.  The scheduling data is ignored.
.TP 8
.B fixpri
The fixed-priority scheduling policy.  CPU usage does not change
the priority of a thread.  The scheduling data is the quantum,
in milliseconds, for the thread.
.PP
\fImpolicy\fR must run as root or a member of group kmem
to gain access to the target objects.
.PP
\fImpolicy\fR interprets the following options:
.TP 8
.B \-host \fImachine\fR
Directs the inquiry to the \fImachid(1)\fR server running
on the specified machine.
.TP 8
.B \-policy \fIpolicy\fR \fIdata\fR
Specifies the new scheduling policy and policy data for
the listed threads.  If other objects, like tasks, processor sets,
or hosts are listed, then \fImpolicy\fR changes the policy
of every thread in the listed objects.
.TP 8
.B \-enable \fIpolicy\fR
Enables the specified scheduling policy for the listed processor sets.
.TP 8
.B \-disable \fIpolicy\fR
Disables the specified scheduling policy for the listed processor sets.
The time-sharing policy may not be disabled.
.TP 8
.B \-threads
When disabling a scheduling policy, specifies that the processor
sets's threads which are using the disabled policy should be changed
to the time-sharing policy.
.SH "SEE ALSO"
ms(1), mcreate(1), machid(1), thread_policy(2),
processor_set_policy_enable(2), processor_set_policy_disable(2).


X11/CHANGES
The following changes have been made to this directory since R3:

     o	the configuration files have moved from util/imake.includes/ to config/
     o	sources for imake are now in config/ instead of util/imake/.



etc/mpolicy/mpolicy.c
/* 
 * Mach Operating System
 * Copyright (c) 1991,1990 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 * 
 * CARNEGIE MELLON ALLOWS FREE USE OF THIS SOFTWARE IN ITS 
 * CONDITION.  CARNEGIE MELLON DISCLAIMS ANY LIABILITY OF ANY KIND FOR
 * ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
 * 
 * Carnegie Mellon requests users of this software to return to
 * 
 *  Software Distribution Coordinator  or  Software.Distribution@CS.CMU.EDU
 *  School of Computer Science
 *  Carnegie Mellon University
 *  Pittsburgh PA 15213-3890
 * 
 * any improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 */
/*
 * HISTORY
 * $Log:	mpolicy.c,v $
 * Revision 2.5  91/08/29  15:48:49  rpd
 * 	Moved machid include files into the standard include directory.
 * 	[91/08/29            rpd]
 * 
 * Revision 2.4  91/03/27  17:27:20  mrt
 * 	Changed mach.h include
 * 
 * Revision 2.3  91/03/19  12:31:43  mrt
 * 	Changed to new copyright
 * 
 * Revision 2.2  90/10/08  13:15:48  rpd
 * 	Created.
 * 	[90/10/08  13:12:33  rpd]
 * 
 */

#include <stdio.h>
#include <strings.h>
#include <mach.h>
#include <mach_error.h>
#include <servers/netname.h>
#include <servers/machid.h>

#define streql(a, b)	(strcmp((a), (b)) == 0)

static mach_port_t server;
static mach_port_t auth;

static void
usage()
{
    quit(1, "usage: mpolicy [-host machine] [-policy policy data] [-enable policy] [-disable policy] [-threads] ids...\n");
}

static int
parse_policy(str)
    char *str;
{
    if (streql(str, "tshare"))
	return POLICY_TIMESHARE;
    else if (streql(str, "fixpri"))
	return POLICY_FIXEDPRI;
    else
	usage();
}

static int
parse_data(policy, str)
    int policy;
    char *str;
{
    return atoi(str);
}

main(argc, argv)
    int argc;
    char *argv[];
{
    char *hostname = "";
    enum { None, ThreadPolicy, PolicyEnable, PolicyDisable } action = None;
    boolean_t ChangeThreads = FALSE;
    int policy, data;
    kern_return_t kr;
    int i, j;

    for (i = 1; i < argc; i++)
	if (streql(argv[i], "-host") && (i < argc-1))
	    hostname = argv[++i];
	else if (streql(argv[i], "-policy") &&
		 (action == None) && (i < argc-2)) {
	    action = ThreadPolicy;
	    policy = parse_policy(argv[++i]);
	    data = parse_data(policy, argv[++i]);
	} else if (streql(argv[i], "-enable") &&
		   (action == None) && (i < argc-1)) {
	    action = PolicyEnable;
	    policy = parse_policy(argv[++i]);
	} else if (streql(argv[i], "-disable") &&
		   (action == None) && (i < argc-1)) {
	    action = PolicyDisable;
	    policy = parse_policy(argv[++i]);
	} else if (streql(argv[i], "-threads"))
	    ChangeThreads = TRUE;
	else if (streql(argv[i], "--")) {
	    i++;
	    break;
	} else if (argv[i][0] == '-')
	    usage();
	else
	    break;

    argv += i;
    argc -= i;

    if ((ChangeThreads && (action != PolicyDisable)) ||
	(action == None))
	usage();

    kr = netname_look_up(name_server_port, hostname, "MachID", &server);
    if (kr != KERN_SUCCESS)
	quit(1, "mpolicy: netname_lookup_up(MachID): %s\n",
	     mach_error_string(kr));

    auth = mach_host_priv_self();
    if (auth == MACH_PORT_NULL)
	auth = mach_task_self();

    for (i = 0; i < argc; i++) {
	mach_id_t id = atoi(argv[i]), origid = id;
	mach_type_t type;

	kr = machid_mach_type(server, auth, id, &type);
	if (kr != KERN_SUCCESS)
	    quit(1, "mpolicy: machid_mach_type: %s\n", mach_error_string(kr));

	if (action == ThreadPolicy) {
	    mthread_t *threads;
	    unsigned int threadCnt;

	    switch (type) {
	      case MACH_TYPE_NONE:
		continue;

	      default:
		kr = machid_mach_lookup(server, auth, id,
					MACH_TYPE_THREAD, &id);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_mach_lookup: %s\n",
			 mach_error_string(kr));

		if (id == 0)
		    goto badtype;
		/* fall-through */

	      case MACH_TYPE_THREAD:
		kr = machid_thread_policy(server, auth, id, policy, data);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_thread_policy: %s\n",
			 mach_error_string(kr));
		continue;

	      case MACH_TYPE_TASK:
		kr = machid_task_threads(server, auth, id,
					 &threads, &threadCnt);
		break;

	      case MACH_TYPE_PROCESSOR_SET_NAME:
		kr = machid_mach_lookup(server, auth, id,
					MACH_TYPE_PROCESSOR_SET, &id);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_mach_lookup: %s\n",
			 mach_error_string(kr));

		if (id == 0)
		    goto badtype;
		/* fall-through */

	      case MACH_TYPE_PROCESSOR_SET:
		kr = machid_processor_set_threads(server, auth, id,
						  &threads, &threadCnt);
		break;

	      case MACH_TYPE_HOST:
		kr = machid_mach_lookup(server, auth, id,
					MACH_TYPE_HOST_PRIV, &id);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_mach_lookup: %s\n",
			 mach_error_string(kr));

		if (id == 0)
		    goto badtype;
		/* fall-through */

	      case MACH_TYPE_HOST_PRIV:
		kr = machid_host_threads(server,auth, id,
					 &threads, &threadCnt);
		break;

	      badtype:
		fprintf(stderr, "mpolicy: %u has type %s\n",
			origid, mach_type_string(type));
		continue;
	    }
	    if (kr != KERN_SUCCESS)
		continue;

	    for (j = 0; j < threadCnt; j++) {
		kr = machid_thread_policy(server, auth,
					  threads[j], policy, data);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_thread_policy: %s\n",
			 mach_error_string(kr));
	    }

	    kr = vm_deallocate(mach_task_self(), (vm_offset_t) threads,
			       (vm_size_t) (threadCnt * sizeof *threads));
	    if (kr != KERN_SUCCESS)
		quit(1, "mpolicy: vm_deallocate: %s\n", mach_error_string(kr));
	} else {
	    switch (type) {
	      case MACH_TYPE_NONE:
		continue;

	      default:
	      case MACH_TYPE_PROCESSOR_SET_NAME:
		kr = machid_mach_lookup(server, auth, id,
					MACH_TYPE_PROCESSOR_SET, &id);
		if (kr != KERN_SUCCESS)
		    quit(1, "mpolicy: machid_mach_lookup: %s\n",
			 mach_error_string(kr));

		if (id == 0)
		    goto badtype;
		/* fall-through */

	      case MACH_TYPE_PROCESSOR_SET:
		switch (action) {
		  case PolicyEnable:
		    kr = machid_processor_set_policy_enable(server, auth,
						id, policy);
		    if (kr != KERN_SUCCESS)
			quit(1, "mpolicy: machid_processor_set_policy_enable: %s\n",
			     mach_error_string(kr));
		    break;

		  case PolicyDisable:
		    kr = machid_processor_set_policy_disable(server, auth,
						id, policy, ChangeThreads);
		    if (kr != KERN_SUCCESS)
			quit(1, "mpolicy: machid_processor_set_policy_disable: %s\n",
			     mach_error_string(kr));
		    break;
		}
		break;
	    }
	}
    }

    exit(0);
}
